experiments/yarn/src/main/scala/io/gearpump/experiments/yarn/client/Client.scala

```diff
@@ -17,323 +17,171 @@
-    system.scheduler.scheduleOnce(Duration(5, TimeUnit.SECONDS)) {
-      LOG.info(s"host=${trackingURL.getHost} port=${trackingURL.getPort} uri=/proxy/${report.getApplicationId}/$restURI/master")
-      lazy val trackingURLConnectionFlow: Flow[HttpRequest, HttpResponse, Any] =
-        Http().outgoingConnection(trackingURL.getHost, trackingURL.getPort)
-      val response = Source.single(RequestBuilding.Get(s"/proxy/${report.getApplicationId}/$restURI/master")).via(trackingURLConnectionFlow).runWith(Sink.head)
-      response.foreach(response => {
-        response.status match {
-          case StatusCodes.OK =>
-            LOG.info(s"status code=${StatusCodes.OK.intValue}")
-            Unmarshal(response.entity).to[String].onComplete(result => {
-              result match {
-                case Success(data) =>
-                  val masterData = read[MasterData](data)
-                  LOG.info(s"leader=${masterData.masterDescription.leader._1}:${masterData.masterDescription.leader._2}")
-                  val cluster=masterData.masterDescription.cluster.map(p=>{p._1+":"+p._2}).mkString(",")
-                  LOG.info("masters=" + cluster)
-                case Failure(throwable) =>
-                  LOG.error("Failed to fetch masters", throwable)
-              }
-              system.shutdown()
-            })
-          case value =>
-            LOG.error(s"Bad status code=${value.intValue}")
-            system.shutdown()
-        }
-      })
-    }
-  }
+  private def uploadConfigToHDFS(appId: ApplicationId): String = {
+    // will use personal home directory so that it will not conflict with other users
+    val confDir = new Path(fs.getHomeDirectory(), s".gearpump_app${appId.getId}/conf/")
 
-  def start(): Boolean = {
-    Try({
-      yarnClient.init(yarnConf)
-      yarnClient.start()
-      true
-    }) match {
-      case Success(success) =>
-        success
-      case Failure(throwable) =>
-        LOG.error("Failed to start", throwable)
-        false
-    }
-  }
+    // copy config from local to remote.
+    val remoteConfFile = new Path(confDir, "gear.conf")
+    var out = fs.create(remoteConfFile)
+    var writer = new OutputStreamWriter(out)
 
-  def submit(): Try[ApplicationId] = {
-    Try({
-      val appContext = yarnClient.createApplication.getApplicationSubmissionContext
-      appContext.setApplicationName(getEnv(YARNAPPMASTER_NAME))
+    val filterJvmReservedKeys = ClusterConfig.filterOutJvmReservedKeys(akka)
 
-      val containerContext = containerLaunchContext(getCommand)
-      appContext.setAMContainerSpec(containerContext)
-      appContext.setResource(getAMCapability)
-      appContext.setQueue(getEnv(YARNAPPMASTER_QUEUE))
+    writer.write(filterJvmReservedKeys.root().render())
+    writer.close()
+    out.close()
 
-      yarnClient.submitApplication(appContext)
-      appContext.getApplicationId
-    })
-  }
+    // save yarn-site.xml to remote
```

Comment:
@huafengw @clockfly @TomaszT  I think this means that when launching the containers (masters, workers) this client YARN file is used rather than what exists on the target container node? Is the assumption that the Gearpump yarnclient will always be launched in the same YARN cluster (cdh version, etc) as where the master and workers will run? @karol-brejna-i is this a safe assumption? Will the Gearpump broker always launch the yarnclient in this environment? In any case it would be reasonable to have "yarn-site.xml" defined as a constant in io.gearpump.experiments.yarn.Constants even though it is not directly referenced by master or workers when they are launched (but i think the conf directory is). In the Constants.scala file we could also note where yarn-site.xml is loaded from and where it is used.